src/code/z_quake.c

```diff
@@ -40,230 +33,257 @@ typedef enum {
 
-        Quake_UpdateShakeInfo(req, shake, perpendicularPertubation, perpendicularPertubation);
-        req->countdown--;
+        Quake_UpdateShakeInfo(req, shake, xyOffset, xyOffset);
+        req->timer--;
     }
-    return req->countdown;
+    return req->timer;
 }
 
 s16 Quake_CallbackType2(QuakeRequest* req, ShakeInfo* shake) {
-    if (req->countdown > 0) {
-        f32 perpendicularPertubation = Quake_Random();
+    if (req->timer > 0) {
+        f32 xyOffset = Quake_Random();
 
-        Quake_UpdateShakeInfo(req, shake, perpendicularPertubation, Rand_ZeroOne() * perpendicularPertubation);
-        req->countdown--;
+        Quake_UpdateShakeInfo(req, shake, xyOffset, Rand_ZeroOne() * xyOffset);
+        req->timer--;
     }
-    return req->countdown;
+    return req->timer;
 }
 
 s16 Quake_CallbackType4(QuakeRequest* req, ShakeInfo* shake) {
-    if (req->countdown > 0) {
-        f32 perpendicularPertubation = Quake_Random() * ((f32)req->countdown / (f32)req->countdownMax);
+    if (req->timer > 0) {
+        f32 xyOffset = Quake_Random() * ((f32)req->timer / req->duration);
 
-        Quake_UpdateShakeInfo(req, shake, perpendicularPertubation, Rand_ZeroOne() * perpendicularPertubation);
-        req->countdown--;
+        Quake_UpdateShakeInfo(req, shake, xyOffset, Rand_ZeroOne() * xyOffset);
+        req->timer--;
     }
-    return req->countdown;
+    return req->timer;
 }
 
 s16 Quake_GetFreeIndex(void) {
     s32 i;
-    s32 ret = 0;
-    s32 min = 0x10000;
+    s32 index = 0;
+    s32 timerMin = UINT16_MAX + 1; // timer is a short, so start with a value beyond its range
 
-    for (i = 0; i < ARRAY_COUNT(sQuakeRequest); i++) {
-        if (sQuakeRequest[i].type == 0) {
-            ret = i;
+    for (i = 0; i < ARRAY_COUNT(sQuakeRequests); i++) {
+        if (sQuakeRequests[i].type == QUAKE_TYPE_NONE) {
+            index = i;
             break;
         }
 
-        if (sQuakeRequest[i].countdown < min) {
-            min = sQuakeRequest[i].countdown;
-            ret = i;
+        if (timerMin > sQuakeRequests[i].timer) {
+            timerMin = sQuakeRequests[i].timer;
+            index = i;
         }
     }
 
-    return ret;
+    return index;
 }
 
-QuakeRequest* Quake_AddImpl(Camera* camera, u32 type) {
+QuakeRequest* Quake_RequestImpl(Camera* camera, u32 type) {
     s16 index = Quake_GetFreeIndex();
-    QuakeRequest* req = &sQuakeRequest[index];
+    QuakeRequest* req = &sQuakeRequests[index];
 
     __osMemset(req, 0, sizeof(QuakeRequest));
+
     req->camera = camera;
     req->camId = camera->camId;
     req->type = type;
-    req->isShakePerpendicular = true;
-    req->randIndex = ((s16)(Rand_ZeroOne() * 0x10000) & ~3) + index;
+    req->isRelativeToScreen = true;
+
+    // Add a unique random identifier to the upper bits of the index
+    // The `~3` assumes there are only 4 requests
+    req->index = index + ((s16)(Rand_ZeroOne() * 0x10000) & ~3);
```

Comment:
Should we have a define or an array count or something for the `4` then?

Reply:
Any ideas on how to generalise this? `~QUAKES_MAX` would quickly fall apart for non-powers of two

Reply:
Yeah, problem is you basically can't... hmm, I suppose you'd have to round up, but there isn't a nice way to do that I don't think.